### PR TITLE
remove try-catch to significantly improve performance

### PR DESCRIPTION
**Pull request by dominictarr**
I've been benchmarking the various bloomfilter implementations available on npm, and this one is very good!

However, the try-catch in `checkEntry` is not necessary, (and also harder for the v8 engine to optimize) 
and removing it makes this module the fastest bloomfilter available in node, even compared to CPP implementations!

below are the results of my benchmark:
```
module-name, add, +test, -test
jsbloom_notrycatch, 219780, 225225, 238663 ## this PR
jsbloom, 171233, 230947, 238095
bloom-filter-cpp, 218818, 210970, 235294
bloomfilter, 170068, 170068, 173310
bloom-lite, 77151, 114811, 103520
bloem, 72493, 76923, 78218
bloom-filter-js, 16441, 16865, 16346
bloomxx, 556, 564, 5372
```
the benchmark was run on a lenovo x201

my bench mark code is up at https://github.com/dominictarr/bloom-bench

### CODE REVIEW

**Comment by dominictarr** [comment]
Oh, then I noticed that `getIndices` returned an array. I tried rewriting that so that it calculates the indices without having to create an array. In the case of `checkEntry` you can return immediately if the entry is missing, and doing so saves you the time of calculating the other hashes on the negative test.

```
module-name, add, +test, -test
jsbloom_notrycatch, 226244, 237530, 273973
jsbloom, 160772, 229358, 247525
bloom-filter-cpp, 220264, 222717, 223214
bloomfilter, 176056, 178571, 181159
bloom-lite, 78685, 125786, 107991
bloem, 77689, 81511, 83416
bloom-filter-js, 17613, 18609, 17751
bloomxx, 644, 564, 5420
```
It's probably possible to get this even faster by optimizing for v8 performance features, but it doesn't really matter, and would require a more scientific benchmark, to measure statistical significance, etc.

**Comment by dominictarr** [comment]
oh, and I was also unsure whether my changes where correct, so I added another test case, to test true negatives. (since the probability setting in the test is so large, I did not actually see any false positives, but allowed 10 just so that the tests should nearly always pass)

File: bloom.js [hunk 1]
```diff
@@ -37,43 +37,30 @@ JSBloom.filter = function(items, target_prob) {
...
 
     addEntry = function(str) {
 
-            var indices = hashes.getIndices(str);
-
-            for (var i = indices.length - 1; i >= 0; i--) {
+        var h1 = hashes.djb2(str)
+        var h2 = hashes.sdbm(str)
+        for (var round = 0; round <= HASH_ROUNDS; round++) {
+            var new_hash = round == 0 ? h1
```
Comment (cry) [inline bloom.js:48]: Personally I would rather do it as it was before due to the more explicit nature, but this is fine. I prefer being able to see what the code is doing at a glance.

File: bloom.js [hunk 2]
```diff
@@ -87,12 +74,32 @@ JSBloom.filter = function(items, target_prob) {
     }
 
     checkEntry = function(str) {
+        var index, extra_indices
```
Comment (cry) [inline bloom.js:77]: Pedantic but keeping with the existing code format would be best (terminate each line with ;).

**Comment by cry** [comment]
Thank you for contributing, I appreciate the changes you've submitted to better jsbloom :)

File: bloom.js (hunk 1 above)
Comment (dominictarr) [reply bloom.js:48]: Yes, definitely. It's unfortunate that the fastest code is often not the prettiest code.

**Comment by cry** [comment]
@dominictarr I am curious though as to why my implementation is the fastest. From what I can see, most of the other fast bloom filters use the same 2 hash optimization trick so shouldn't these filters be roughly the same speed?

**Comment by dominictarr** [comment]
Yes, https://github.com/jasondavies/bloomfilter.js and https://github.com/bbondy/bloom-filter-js are also in a similar ball park. Once the algorithm is correct, it becomes about optimizing the javascript, to be the kind of thing that js engines can optimize effectively.
I learnt a bit about that by implementing https://github.com/crypto-browserify/sha.js (although, other people have since improved it much more than I was able to on my own)

try-catch is one of the things which v8 at least, doesn't try to optimize. Creating an array, then throwing it away is also slower than not creating an array. Assigning to an array is also slower than a local variable.
But, you can't really apply general rules, you have a take a guess and then measure, and repeat.

So, it was probably just a coincidence that yours was the fastest, but your code was pretty direct, and happened to use patterns that are easy for js engines.

**Pull request merged**